### DISCUSSION TITLE
How to map nested arrays in thymeleaf

### DISCUSSION THREAD

**Discussion by pellerito** [thymeleaf/thymeleaf#1017] [Q&A]
This is my code:

My classes:

`public class Persons {

    List<Student> students;
}
public class Student {

    List<String> role;
}`

In thymeleaf:

`<select id="role" tabindex="16" class="form-select input-widget custom-size" th:field="*{students[0].role[0]}">
    <option value="1">Mathematician</option>
    <option value="2">Economist</option>                                                                          
</select>`

It means that every student can have multiple roles. The issue is that the program iterate only the first array (i.e. students[1].role[0], students[2].role[0]), but not the second array. Is there a way to control which array will iterated? I want that in this select tag only the second array is iterated. My aim is that the role is added to the role list.

I tried the above example, but only the first array is iterated.

Info: With preprocessing it doesnt work, because the items arent already added to the inner array.

**Comment by vivek656** [comment]
you can use nested <th:each>  .
check nullability for role as well.
``` 
<th:block th:each="student: *{students}">
        <th:block th:each="role=${student.role}"></th:block>     
</th:block>
```

**Comment by conorheffron** [comment]
@pellerito I think you need to update thymeleaf template with`nested table` per each table data segment (`td`).
 - In your case each student can have multiple roles, see sample below where each person can have many employers (in their job history).


I have a sample app running where _nested list of employers per person_ is rendered via `Thymeleaf templates` (against Java Spring Boot Server). 

Let me know if this is useful (see details below).

## Deployed at: http://34.147.243.27:8080/
## Code Repository: https://github.com/conorheffron/ironoc-db

---
#### Thymeleaf (Frontend)
 - https://github.com/conorheffron/ironoc-db/blob/main/src/main/resources/templates/employee-list.html
```
<tbody>
                    <tr th:each="person : ${personsList}">
                        <td th:text="${person.id}"></td>
                        <td th:text="${person.title}"></td>
                        <td th:text="${person.firstName}"></td>
                        <td th:text="${person.surname}"></td>
                        <td th:text="${person.age}"></td>
.
.
.
.
                        <td>
                            <table class="table table-sm mb-0">
                                <thead>
                                <tr>
                                    <th>Job Title</th>
                                    <th>Employer Name</th>
                                    <th>Start Year</th>
                                </tr>
                                </thead>
                                <tbody>
                                <tr th:each="employer : ${person.employers}">
                                    <td th:text="${employer.title}"></td>
                                    <td th:text="${employer.employerName}"></td>
                                    <td th:text="${employer.startYear}"></td>
                                </tr>
                                </tbody>
                            </table>
                        </td>
                    </tr>
.
.
.
                    </tbody>
```

#### Java Spring Boot (Backend)
 - https://github.com/conorheffron/ironoc-db/blob/main/src/main/java/com/ironoc/db/controller/PersonController.java -> **Controller for GET mapping** (which uses JPA CrudRepository call for SQL DB which includes findAll method by default)
 - https://github.com/conorheffron/ironoc-db/blob/main/src/main/java/com/ironoc/db/model/Person.java (**List of nested employer POJOs**)
 - https://github.com/conorheffron/ironoc-db/blob/main/src/main/java/com/ironoc/db/service/PersonServiceImpl.java (**service with findAll call**)
 - https://github.com/conorheffron/ironoc-db/blob/main/src/main/java/com/ironoc/db/dao/PersonDao.java (**CrudRepository**)
```
@GetMapping(value = "/")
	public String home(ModelMap map) {
		log.info("Entering personController.home: map={}", map);
		
		List<Person> personslist = personService.getAllPersons();
        map.addAttribute("personsList", personslist);
        map.addAttribute("person", Person.builder().build());

        return "index";
	}
```
---
## Rendered Output (`Job History` column has nested list of employer details)
<img width="1664" height="1172" alt="image" src="https://github.com/user-attachments/assets/77d21f07-28be-44fd-8c5a-b8bcfafbd101" />
